### DISCUSSION TITLE
Is it possible to boot from SD on SSD202?

### DISCUSSION THREAD

**Discussion by ltrsunny** [linux-chenxing/linux-chenxing.org#43] [General]
Does the SoC only try to find bootloader from flash?

**Comment by fifteenhex** [comment]
Scratching my memory here.. I don't think so. There's no accesses to the SD register from the bootrom. I think the second stage loader can but that means you always need SPI NOR or SPI NAND to have the IPL.

**Reply by rperier** [reply]
Once we have sd card support mainline,  think about kexec and petitboot for  example (even if it does not reply completly to the question) ;)

**Reply by ltrsunny** [reply]
Well... what an annoying chip. I am thinking about how to replace the kernel & rootfs from SOM vendor. 

Does the SPI NAND work faster than SD port? If not, I think petiboot is a better choice for my production.

**Reply by fifteenhex** [reply]
You want to boot from local storage without nuking the vendor stuff?
If you can get a kernel to load in their u-boot then you can put your rootfs on USB or SD.

**Reply by ltrsunny** [reply]
> You want to boot from local storage without nuking the vendor stuff? If you can get a kernel to load in their u-boot then you can put your rootfs on USB or SD.

Yes. Let me try to compile my kernel first...

**Reply by fifteenhex** [reply]
Ok, to get the mainline-ish kernel to boot with their u-boot you need to do something like:

```
mw 0x16002000 0x1e0; setenv serverip 192.168.3.235; setenv ipaddr 192.168.3.211; setenv loadaddr 0x22000000; estart; tftpboot kernel_sbc2d70; go ${loadaddr}
```

You need to have the option for a DTB appended to the end of the kernel and your DTB on the end of your kernel image.

**Reply by ltrsunny** [reply]
emmm……I tried to build the code from 'mstar_v5_15_rebase' branch @ /linux-chenxing/linux, and got many errors at 'drivers/pinctrl/mstar/pinctrl-mstar.c'; tried the code from linux mainline, build succeed but cannot boot...
I think I should try the ones you compiled first...

**Reply by ltrsunny** [reply]
It seems like ‘after “go ${loadaddr}”,U-boot loaded the dtb and start the normal boot to the vendor's kernel in nand’.

**Reply by fifteenhex** [reply]
> It seems like ‘after “go ${loadaddr}”,U-boot loaded the dtb and start the normal boot to the vendor's kernel in nand’.

Can you paste all of the u-boot output? It'll help to see what is happening.
Also, which board are you trying to boot on? I can make you a known working kernel image.

**Reply by ltrsunny** [reply]
> Can you paste all of the u-boot output? It'll help to see what is happening.
> Also, which board are you trying to boot on? I can make you a known working kernel image.

It‘s changed after a cold boot. Right now I am following your tutorial from [here](https://github.com/fifteenhex/buildroot_idosom2d01#readme). 

The command 'rz' locked my board, I am using tftp to transfer(very slow, only about 1KB/s) the image under rescue mode.

What I am using is the SSD202 version of SBC2D06. Thx for your great work.

**Reply by fifteenhex** [reply]
> The command 'rz' locked my board, I am using tftp to transfer(very slow, only about 1KB/s) the image under rescue mode.

Try setting the block size to 4096 or bigger.

**Reply by ltrsunny** [reply]
Let me try… I've just installed github's app, but still cannot get notification of discussion on time😂

**Reply by fifteenhex** [reply]
I think @ddB0515 recently did the flashing stuff and had a few issues with my instructions.
If you look through the closed issues and the closed PR he created there might be some useful information if you get stuck.

**Reply by ddB0515** [reply]
@ltrsunny check my [IDO-SOM2D01-Flashing.md](https://github.com/ddB0515/buildroot_idosom2d01/blob/flashing-som2d01/IDO-SOM2D01-Flashing.md) maybe you will find something useful 
For tftp I had same issue but resolved with higher block like @fifteenhex mention it
`tftp -g -r kernel.fit 192.168.1.xxx -b 65464`

**Reply by ltrsunny** [reply]
> @ltrsunny check my [IDO-SOM2D01-Flashing.md](https://github.com/ddB0515/buildroot_idosom2d01/blob/flashing-som2d01/IDO-SOM2D01-Flashing.md) maybe you will find something useful 
> For tftp I had same issue but resolved with higher block like @fifteenhex mention it
> `tftp -g -r kernel.fit 192.168.1.xxx -b 65464`

Yes, I already read about it and working on it now. Thank you.

**Reply by ltrsunny** [reply]
All images burned in.
After  
```
setenv bootargs console=ttyS0,115200 ubi.mtd=1 rootfstype=ubifs root=ubi0:rootfs rw clk_ignore_unused init=linuxrc
setenv bootcmd 'loadaddr 0x22000000; ubi readvol ${loadaddr} kernel 0x800000; bootm ${loadaddr}'
saveenv
```
got an
```
Hit any key to stop autoboot:  0
Unknown command 'loadaddr' - try 'help'
Volume 0x800000 not found!
Wrong Image Format for bootm command
ERROR: can't get kernel image!
```

Is there something missing in this version?
```
=> version
U-Boot 2021.01-rc4 (Nov 06 2021 - 15:45:25 +0000)

arm-buildroot-linux-gnueabihf-gcc.br_real (Buildroot 2021.08-1184-gd9123dd799) 11.2.0
GNU ld (GNU Binutils) 2.37
```
```
=> loadaddr
Unknown command 'loadaddr' - try 'help'

```

@fifteenhex @ddB0515  Anyway, it is already a big step for me, thanks a lot.

**Reply by fifteenhex** [reply]
I think `'loadaddr 0x22000000` should be `'setenv loadaddr 0x22000000`

**Reply by ddB0515** [reply]
correct, I have typo to fix :)

**Reply by ltrsunny** [reply]
YES! It works now.

**Reply by fifteenhex** [reply]
That's great news. Thanks for sticking with it.

**Reply by ltrsunny** [reply]
Emm... still something wrong with rootfs partition, but I think I can use mine on SD card directly not.

**Reply by ltrsunny** [reply]
> That's great news. Thanks for sticking with it.

Thanks for your fancy work.

**Reply by ddB0515** [reply]
My tutorial was written for rootfs with UBIFS so try to make it and flash it (should work fine) as I didn't try with squashfs :(

**Reply by fifteenhex** [reply]
I guess you're getting "failed to mount root blah blah"?

If you used the binaries from the github repo then it doesn't support the ubifs root like @ddB0515 's instructions have.

A rootfs on SD card or USB should work

**Reply by ltrsunny** [reply]
> My tutorial was written for rootfs with UBIFS so try to make it and flash it (should work fine) as I didn't try with squashfs :(

That should be the key.

**Reply by fifteenhex** [reply]
Going to add the command line to boot the current images to the readme now..

**Reply by ltrsunny** [reply]
There are some looping information during boot like, are they from the vendor's uboot-spl which I did not erased completely？
```
mtd read ret = 0, rlen 64
b: 2f1, o: 800, d: 2380e4dc, l: 40
w: 13
w: 00
w: bc
w: 41
w: 0f
w: c0
w: 0b
w: 00
w: 00
w: ff
mtd read ret = 0, rlen 64
b: 2f2, o: 800, d: 2380e51c, l: 40
w: 13
w: 00
w: bc
w: 81
w: 0f
w: c0
w: 0b
w: 00
w: 00
w: ff
```

**Reply by fifteenhex** [reply]
No, that's from my/our u-boot. There is some issue with the SPI NAND not working correctly. So that's some debug info.
Taking that printing out stops it loading properly. When I get some time I'll fix it and it'll go away.

**Reply by ltrsunny** [reply]
> No, that's from my/our u-boot. There is some issue with the SPI NAND not working correctly. So that's some debug info. Taking that printing out stops it loading properly. When I get some time I'll fix it and it'll go away.

Okay thanks.

**Reply by fifteenhex** [reply]
Try this:

```
setenv bootargs console=ttyS0,115200 ubi.mtd=1 ubi.block=0,4 root=/dev/ubiblock0_4 clk_ignore_unused
setenv bootcmd 'setenv loadaddr 0x22000000; ubi readvol ${loadaddr} kernel 0x800000; bootm ${loadaddr}#ssd202d-sbc2d06
```

**Reply by ltrsunny** [reply]
> Try this:
> 
> ```
> setenv bootargs console=ttyS0,115200 ubi.mtd=1 ubi.block=0,4 root=/dev/ubiblock0_4 clk_ignore_unused
> setenv bootcmd 'setenv loadaddr 0x22000000; ubi readvol ${loadaddr} kernel 0x800000; bootm ${loadaddr}#ssd202d-sbc2d06
> ```

It booted! Thanks again for your great help.  お疲れ様でした！

I think I will try to boot from SD tomorrow and keep learning about kernel compilation.

**Reply by ltrsunny** [reply]
`setenv bootargs console=ttyS0,115200 root=/dev/mmcblk0 rootwait rw clk_ignore_unused` 

The arguments worked, but

```
[    4.383044] systemd[1]: System time before build time, advancing clock.
[    4.418155] systemd[1]: Failed to look up module alias 'autofs4': Function not implemented
[    4.542503] systemd[1]: systemd 248.3-1ubuntu8 running in system mode. (+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS -OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP -LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
[    4.574136] systemd[1]: Detected architecture arm.

Welcome to Ubuntu 21.10!

[    4.629280] systemd[1]: Hostname set to <ubuntu>.
[    4.707706] systemd[1]: Failed to create control group inotify object: Function not implemented
[    4.737498] systemd[1]: Failed to allocate manager object: Function not implemented
[!!!!!!] Failed to allocate manager object.
[    4.777912] systemd[1]: Freezing execution.
[   33.127812] vcc_dram: disabling
[   91.207884] random: crng init done
```

It looks some kernel options missed in compilation, I think I should try to build one by myself.

So is ‘idosom2d01-kernel.fit’ a uImage with dtb in the end? @fifteenhex

**Reply by fifteenhex** [reply]
> [    4.383044] systemd[1]: System time before build time, advancing clock.
> [    4.418155] systemd[1]: Failed to look up module alias 'autofs4': Function not implemented
> [    4.542503] systemd[1]: systemd 248.3-1ubuntu8 running in system mode. (+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS -OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP -LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
> [    4.574136] systemd[1]: Detected architecture arm.
> 
> Welcome to Ubuntu 21.10!

Probably the first time ubuntu has ever run on these chips. Long term I think you'll want to customise buildroot as ubuntu is going to eat a lot of the tiny ram.

> [    4.629280] systemd[1]: Hostname set to <ubuntu>.
> [    4.707706] systemd[1]: Failed to create control group inotify object: Function not implemented
> [    4.737498] systemd[1]: Failed to allocate manager object: Function not implemented
> [!!!!!!] Failed to allocate manager object.
> [    4.777912] systemd[1]: Freezing execution.
> [   33.127812] vcc_dram: disabling
> [   91.207884] random: crng init done
> ```
> 
> It looks some kernel options missed in compilation, I think I should try to build one by myself.

You need to enable inotify and dnotify in the filesystems menu of the kernel config.
Checking out the buildroot repo, letting it build for an hour while you watch youtube, then using `make linux-config` to change the kernel config and then running `make` again to get it to regenerate the kernel image for you is probably the simplest way to go forward.
If you get that working you can then reuse it to checkout the kernel code locally and start hacking on it. I use the toolchain etc from the buildroot to build my development images.

> 
> So is ‘idosom2d01-kernel.fit’ a uImage with dtb in the end? @fifteenhex

Nope. The images from the buildroot setup are "flattened image trees". They have a kernel, maybe a ramdisk, and all of the DTBs for all of the different types of board. That means the single file can boot on any ssd202d board we find.

Cheers,

Daniel

**Reply by ltrsunny** [reply]
Thanks Daniel.
> Probably the first time ubuntu has ever run on these chips. Long term I think you'll want to customise buildroot as ubuntu is going to eat a lot of the tiny ram.

I also tried Debian on it, got the same errors. Both of them were running on Allwinner V3s which is much weaker than SSD202. I think a common distro would be a better choice for me, because later I need to implement more functions in this system.



> You need to enable inotify and dnotify in the filesystems menu of the kernel config.
> Checking out the buildroot repo, letting it build for an hour while you watch youtube, then using `make linux-config` to change the kernel config and then running `make` again to get it to regenerate the kernel image for you is probably the simplest way to go forward.
> If you get that working you can then reuse it to checkout the kernel code locally and start hacking on it. I use the toolchain etc from the buildroot to build my development images.

Ok let me try first.

**Reply by ltrsunny** [reply]
> You need to enable inotify and dnotify in the filesystems menu of the kernel config.
> Checking out the buildroot repo, letting it build for an hour while you watch youtube, then using `make linux-config` to change the kernel config and then running `make` again to get it to regenerate the kernel image for you is probably the simplest way to go forward.
> If you get that working you can then reuse it to checkout the kernel code locally and start hacking on it. I use the toolchain etc from the buildroot to build my development images.

I cloned /buildroot_idosom2d01, but 'No rule to make target linux-config', is this an option after once build?

**Reply by fifteenhex** [reply]
Sorry it's `make buildroot-linux-menuconfig`

**Reply by ltrsunny** [reply]
Got it.

**Reply by ltrsunny** [reply]
> Ok, to get the mainline-ish kernel to boot with their u-boot you need to do something like:
> 
> ```
> mw 0x16002000 0x1e0; setenv serverip 192.168.3.235; setenv ipaddr 192.168.3.211; setenv loadaddr 0x22000000; estart; tftpboot kernel_sbc2d70; go ${loadaddr}
> ```
> 
> You need to have the option for a DTB appended to the end of the kernel and your DTB on the end of your kernel image.

A small question, what is the purpose of `mw 0x16002000 0x1e0`

**Comment by ltrsunny** [comment]
Some updates:

Never succeed building kernel via buildroot.
Try to use the code from mstar_v5_15_rebase (5_16_rebase also failed) to get the uImage and dtb.
```
mw 0x16002000 0x1e0
dhcp 0x22000000 192.168.50.211:uImage
dhcp 0x22800000 192.168.50.211:mstar-infinity2m-ssd202d-wirelesstag-ido-sbc2d06-v1b-22w.dtb
bootm 0x22000000 - 0x22800000
```
and froze at
```
=> bootm 0x22000000 - 0x22800000
## Booting kernel from Legacy Image at 22000000 ...
   Image Name:   Linux-5.15.012131433+
   Image Type:   ARM Linux Kernel Image (uncompressed)
   Data Size:    5037112 Bytes = 4.8 MiB
   Load Address: 21800000
   Entry Point:  21800000
   Verifying Checksum ... OK
## Flattened Device Tree blob at 22800000
   Booting using the fdt blob at 0x22800000
   Loading Kernel Image
   Loading Device Tree to 26f7d000, end 26f863dd ... OK

Starting kernel ...


```
Is it about wrong entry point or wrong kernel configuration?

**Reply by fifteenhex** [reply]
Can you try adding CONFIG_DEBUG_MSTARV7_PMUART to you config?
That should print out very early what is going wrong.

**Reply by ltrsunny** [reply]
![image](https://user-images.githubusercontent.com/49466661/146012694-fb761202-3dad-4a00-b8f8-f00507112c9c.png)
Which port should I use？

**Reply by fifteenhex** [reply]
Ok, that'll be the problem. You need to select `CONFIG_ARCH_MSTARV7`.
I guess we should send a patch to get enabled in the ARM v7 multiconfig... I'll add that to my tree now.

**Reply by ltrsunny** [reply]
Wait… my config was changed by something, ARCH_MSTARV7 became n.

**Reply by ltrsunny** [reply]
> Can you try adding CONFIG_DEBUG_MSTARV7_PMUART to you config? That should print out very early what is going wrong.

Enabled but nothing output.

**Reply by ltrsunny** [reply]
Does the kernel's compressed format affect? I used LZ4, I saw yours is XZ, does your u-boot configuration support? @fifteenhex

**Reply by ltrsunny** [reply]
> Does the kernel's compressed format affect? I used LZ4, I saw yours is XZ, does your u-boot configuration support? @fifteenhex

Tried XZ, doesn't work…
```

=> bootm 0x22000000 - 0x22800000
## Booting kernel from Legacy Image at 22000000 ...
   Image Name:   Linux-5.15.012142315+
   Image Type:   ARM Linux Kernel Image (uncompressed)
   Data Size:    3092376 Bytes = 2.9 MiB
   Load Address: 21800000
   Entry Point:  21800000
   Verifying Checksum ... OK
## Flattened Device Tree blob at 22800000
   Booting using the fdt blob at 0x22800000
   Loading Kernel Image
   Loading Device Tree to 26f7d000, end 26f863dd ... OK

Starting kernel ...


```

**Reply by fifteenhex** [reply]
Compression shouldn't matter as that happens after u-boot has finished. I suspect the decompression is failing. There is an option to enable the decompressor output in the debug menu that might help: `CONFIG_DEBUG_UNCOMPRESS`.
Can you try that? you should at least get a message about decompressing the kernel.

**Reply by ltrsunny** [reply]
Let me try in the morning. Thanks a lot.

**Reply by ltrsunny** [reply]
> Compression shouldn't matter as that happens after u-boot has finished. I suspect the decompression is failing. There is an option to enable the decompressor output in the debug menu that might help: `CONFIG_DEBUG_UNCOMPRESS`. Can you try that? you should at least get a message about decompressing the kernel.
```
Starting kernel ...

Uncompressing Linux... done, booting the kernel.

```
then stopped.

I think I should enable more debug options.

**Comment by ltrsunny** [comment]
Reuse your linux.config and booted now, I think I modified some key options on mine. Let me try to address.

**Reply by ltrsunny** [reply]
It is CONFIG_SLUB. 
I selected CONFIG_SLOB before……
